tests/basic_ref.rs

```diff
@@ -0,0 +1,362 @@
+use flurry::HashMap;
+use std::sync::Arc;
+
+#[test]
+fn insert() {
+    let map = HashMap::<usize, usize>::new();
+    let map = map.as_ref();
+
+    let old = map.insert(42, 0);
+    assert!(old.is_none());
+}
+
+#[test]
+fn get_empty() {
+    let map = HashMap::<usize, usize>::new();
+
+    {
+        let map = map.as_ref();
+        let e = map.get(&42);
+        assert!(e.is_none());
+    }
+}
+
+#[test]
+fn get_key_value_empty() {
+    let map = HashMap::<usize, usize>::new();
+
+    {
+        let map = map.as_ref();
+        let e = map.get_key_value(&42);
+        assert!(e.is_none());
+    }
+}
+
+#[test]
+fn remove_empty() {
+    let map = HashMap::<usize, usize>::new();
+
+    {
+        let map = map.as_ref();
+        let old = map.remove(&42);
+        assert!(old.is_none());
+    }
+}
+
+#[test]
+fn insert_and_remove() {
+    let map = HashMap::<usize, usize>::new();
+
+    {
+        let map = map.as_ref();
+        map.insert(42, 0);
+        let old = map.remove(&42).unwrap();
+        assert_eq!(old, &0);
+        assert!(map.get(&42).is_none());
+    }
+}
+
+#[test]
+fn insert_and_get() {
+    let map = HashMap::<usize, usize>::new();
+
+    map.as_ref().insert(42, 0);
+    {
+        let map = map.as_ref();
+        let e = map.get(&42).unwrap();
+        assert_eq!(e, &0);
+    }
+}
+
+#[test]
+fn insert_and_get_key_value() {
+    let map = HashMap::<usize, usize>::new();
+
+    map.as_ref().insert(42, 0);
+    {
+        let map = map.as_ref();
+        let e = map.get_key_value(&42).unwrap();
+        assert_eq!(e, (&42, &0));
+    }
+}
+
+#[test]
+fn update() {
+    let map = HashMap::<usize, usize>::new();
+
+    let map1 = map.as_ref();
+    map1.insert(42, 0);
+    let old = map1.insert(42, 1);
+    assert_eq!(old, Some(&0));
+    {
+        let map2 = map.as_ref();
+        let e = map2.get(&42).unwrap();
+        assert_eq!(e, &1);
+    }
+}
+
+#[test]
+fn concurrent_insert() {
+    let map = Arc::new(HashMap::<usize, usize>::new());
+
+    let map1 = map.clone();
+    let t1 = std::thread::spawn(move || {
+        for i in 0..64 {
+            (*map1).as_ref().insert(i, 0);
+        }
+    });
+    let map2 = map.clone();
+    let t2 = std::thread::spawn(move || {
+        for i in 0..64 {
+            (*map2).as_ref().insert(i, 0);
+        }
+    });
+
+    t1.join().unwrap();
+    t2.join().unwrap();
+
+    let map = (*map).as_ref();
+    for i in 0..64 {
+        let v = map.get(&i).unwrap();
+        assert!(v == &0 || v == &1);
+
+        let kv = map.get_key_value(&i).unwrap();
+        assert!(kv == (&i, &0) || kv == (&i, &1));
+    }
+}
+
+#[test]
+fn concurrent_remove() {
+    let map = Arc::new(HashMap::<usize, usize>::new());
+
+    {
+        let map = (*map).as_ref();
```

Comment:
I'm not thrilled with this myself -- I think it's because `<Arc as AsRef>::as_ref` takes precedence over the deref to our inherent `HashMap::as_ref`. Maybe we need a different name, like `HashMap::guarded()`. A name like that might pair better with a user's guard too, `HashMap::with_guard(guard: &guard)`.